src/Razor/Microsoft.AspNetCore.Razor.Language/src/Components/ComponentWhitespacePass.cs

```diff
@@ -103,12 +133,48 @@ private static void RemoveContiguousWhitespace(IntermediateNodeCollection nodes,
                     break;
                 }
             }
+
+            return countRemoved;
         }
 
         enum TraversalDirection
         {
             Forwards,
             Backwards
         }
+
+        class Visitor : IntermediateNodeWalker
+        {
+            public override void VisitMarkupElement(MarkupElementIntermediateNode node)
+            {
+                RemoveContiguousWhitespace(node.Children, TraversalDirection.Forwards);
+                RemoveContiguousWhitespace(node.Children, TraversalDirection.Backwards);
+                VisitDefault(node);
+            }
+
+            public override void VisitTagHelperBody(TagHelperBodyIntermediateNode node)
+            {
+                // The goal here is to remove leading/trailing whitespace inside component child content. However,
+                // at the time this whitespace pass runs, ComponentChildContent is still TagHelperBody in the tree.
+                RemoveContiguousWhitespace(node.Children, TraversalDirection.Forwards);
+                RemoveContiguousWhitespace(node.Children, TraversalDirection.Backwards);
+                VisitDefault(node);
+            }
+
+            public override void VisitDefault(IntermediateNode node)
+            {
+                // For any CSharpCodeIntermediateNode children, remove their preceding and trailing whitespace
+                for (var childIndex = 0; childIndex < node.Children.Count; childIndex++)
+                {
+                    if (node.Children[childIndex] is CSharpCodeIntermediateNode)
+                    {
+                        childIndex -= RemoveContiguousWhitespace(node.Children, TraversalDirection.Backwards, childIndex - 1);
+                        RemoveContiguousWhitespace(node.Children, TraversalDirection.Forwards, childIndex + 1);
```

Comment:
Trying to understand what is happening here. Why is this restarting the search from wherever it traversed to in the backwards direction? Why not just start it from `childIndex + 1` without doing the `childIndex -=` in the previous line. Could definitely use a comment here to explain.

Reply:
I’ll add a comment. We have to adjust the index here to account for having mutated the collection we’re iterating over.

Reply:
Ah, makes sense